sklearn/model_selection/_classification_threshold.py

```diff
@@ -996,8 +1037,18 @@ def get_metadata_routing(self):
 
     def _get_curve_scorer(self):
         """Get the curve scorer based on the objective metric used."""
-        scoring = check_scoring(self.estimator, scoring=self.scoring)
+        scorer = check_scoring(self.estimator, scoring=self.scoring)
+        # XXX: at the time of writing, there is no very explicit way to check
+        # if a scorer expects thresholded binary classification predictions.
+        # TODO: updates this condition when a better way is available.
+        if scorer._response_method != "predict":
```

Comment:
The following alternative is more verbose but more directly test for the assumption we make about valid scorer objects.

```suggestion
        scorer_response_methods = scorer._response_method
        if isinstance(scorer_response_methods, str):
            scorer_response_methods = {scorer_response_methods}
        else:
            scorer_response_methods = set(scorer_response_methods)

        if scorer_response_methods.issubset(["predict_proba", "decision_function"]):
```

Reply:
This accesses the private `scorer._response_method`. Maybe make it conditional on `hasattr(scorer, "_response_method")`. Then I'm fine.

Also, I do not consider a change in raising errors an API that needs to be deprecated - in case we change this later.

Reply:
@adrinjalali What do you think?

Reply:
I think this logic needs to be moved to the scorer. The scorer should raise if it's trying to apply a classification on a regressor for instance.

Reply:
But here the problem is that we pass thresholded predictions of a classifier (computed by `TunedThresholdClassifierCV` to a scorer that was meant for unthresholded outputs of `decision_function` / `predict_proba`. Since it's the `TunedThresholdClassifierCV` that is in charge of doing the thresholding (via the private curve scorer), and not the estimator, I don't see how to move this logic to the scorer.

Here is what happens in this meta-estimator:

- `TunedThresholdClassifierCV` asks the estimator for unthresholded predictions (from `decision_function` or `predict_proba`;
- `TunedThresholdClassifierCV` transforms those into thresholded predictions for various values of the threshold (as if it had called the `predict` method of the classifier many times with different values of the threshold);
- For each value of the threshold, `TunedThresholdClassifierCV` passes the thresholded predictions to a scorer that should expect thresholded predictions.

The problem here is the last step if `TunedThresholdClassifierCV` passes the thresholded predictions to a scorer that expects unthresholded predictions (ROC AUC, log loss,...), then we have a methodological problem: the result is meaningless to me. The estimator is no longer involved at this stage. 

The problem of classification vs regression is an orthogonal problem and I agree this would be easy to fix in `scorer.score(estimator, X, y)`, but fixing this problem is not the point of this PR.

Reply:
> This accesses the private scorer._response_method. Maybe make it conditional on hasattr(scorer, "_response_method"). Then I'm fine.

I updated the suggestion above to not raise the `ValueError` for `scorer` objects that do not expose the `_response_method` private attribute.